[WebXR][OpenXR] Fill in FrameData information

svillar opened the pull request:
#### a57ebb1d227dc33791da87d0fa7546a88053c98a
<pre>
[WebXR][OpenXR] Fill in FrameData information
<a href="https://bugs.webkit.org/show_bug.cgi?id=295881">https://bugs.webkit.org/show_bug.cgi?id=295881</a>

Reviewed by Carlos Garcia Campos.

XR information coming from XR devices is retrieved in the UI process
using OpenXR API. That info is then later transferred to the Web
Process to be consumed by the WebXR API. A struct called FrameData
is used to carry that information.

In this commit we&apos;re among other things collecting &amp; sending:
- Views information: fov, pose
- Tracking information availability
- Floor location (transform from local space)
- Layer data (so far only 1 projection layer)

In the WebProcess the submitFrame() method receives a vector
of objects containing information about the layers. The WebXR
layers spec is still not available in WebKit but even without
that there is still at least one layer, the base layer.

That info was not passed to the UIProcess in submitFrame() because
it wasn&apos;t used by the ARKit implementation. However it&apos;s needed
for the OpenXR one because we get there the viewport information
that we need to create the stereo views (we use a single projection
layer that is split in as many parts as views we have, typically 2).

Serializing PlatformXR::Device::Layer proved to be really complex
as serializers do not deal well with having multiple nested namespaces
for a given type. That&apos;s why we had to &quot;clone&quot; that structure in
another one hanging directly from the WebKit namespace.

Also a new message had to be added as well to ask the UIProcess to
create the projection layer with the data coming from WebXR.

One of the most important pieces of information to send to the Web Process
is the texture were the WebGL rendering should happen to be later submitted
to the OpenXR compositor. This commit implements the required machinery
to acquire and release those textures coming from OpenXR. However that
cannot be directly used by WebGL, we need to eventually implement a similar
mechanism as the ExternalTexture used by Apple ports. That&apos;s why no
rendering is expected to actually happen in the texture provided by OpenXR.

Last but not least, when adding the swapchains code we detected 2 important
issues in the current implementation:
- OpenXR resources were not properly freed upon session exiting
- we were incorrectly running multiple sequences of xrBeginFrame-&gt;xrEndFrame
in parallel. The confusion arose from an incorrect interpretation of what
xrWaitFrame does. The current code assumed that xrWaitFrame would wait until the
previous xrEndFrame is finished. But that&apos;s wrong. As the specs say
&quot;A subsequent xrWaitFrame call must block until the previous frame has been begun
with xrBeginFrame and must unblock independently of the corresponding call
to xrEndFrame&quot;. That&apos;s why we had to add some additional synchronization
using binary semaphores, basically identical to what ARKit implementation does.

* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/Shared/XR/XRDeviceLayer.h: Copied from Source/WebKit/UIProcess/XR/openxr/PlatformXRSystemOpenXR.cpp.
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::initializeTrackingAndRendering):
(WebKit::XRDeviceProxy::submitFrame):
* Source/WebKit/Shared/XR/XRSystem.serialization.in:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/XR/PlatformXRCoordinator.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::submitFrame):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.messages.in:
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp: Added.
(WebKit::OpenXRLayerProjection::create):
(WebKit::OpenXRLayerProjection::OpenXRLayerProjection):
(WebKit::OpenXRLayerProjection::startFrame):
(WebKit::OpenXRLayerProjection::endFrame):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.h: Added.
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.cpp: Added.
(WebKit::OpenXRSwapchain::create):
(WebKit::OpenXRSwapchain::OpenXRSwapchain):
(WebKit::OpenXRSwapchain::~OpenXRSwapchain):
(WebKit::OpenXRSwapchain::acquireImage):
(WebKit::OpenXRSwapchain::releaseImage):
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.h: Added.
(WebKit::OpenXRSwapchain::swapchain const):
(WebKit::OpenXRSwapchain::width const):
(WebKit::OpenXRSwapchain::height const):
(WebKit::OpenXRSwapchain::size const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
(WebKit::XrIdentityPose):
(WebKit::XrPosefToPose):
(WebKit::XrViewToView):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::~OpenXRCoordinator):
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::createLayerProjection):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::submitFrameInternal):
(WebKit::OpenXRCoordinator::submitFrame):
(WebKit::OpenXRCoordinator::collectViewConfigurations):
(WebKit::OpenXRCoordinator::initializeBlendModes):
(WebKit::OpenXRCoordinator::cleanupSessionAndAssociatedResources):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
(WebKit::OpenXRCoordinator::populateFrameData):
(WebKit::OpenXRCoordinator::createReferenceSpacesIfNeeded):
(WebKit::OpenXRCoordinator::renderLoop):
(WebKit::OpenXRCoordinator::recommendedResolution const): Deleted.
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
* Source/WebKit/UIProcess/XR/openxr/PlatformXRSystemOpenXR.cpp:
(WebKit::PlatformXRSystem::createLayerProjection):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::createLayerProjection):
(WebKit::PlatformXRSystemProxy::submitFrame):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: <a href="https://commits.webkit.org/297780@main">https://commits.webkit.org/297780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb826bad3feefa40ec086fae9bc6bb18218b2bab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112814 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23027 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119017 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63344 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33201 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41112 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85855 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36511 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26497 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101485 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66161 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25786 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19616 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62776 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95890 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122238 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39892 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29738 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94716 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40275 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97705 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94454 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24116 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39581 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17395 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35995 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39778 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45276 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39417 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42751 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41156 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/XR/PlatformXRSystem.messages.in:
```diff
@@ -37,7 +37,13 @@ messages -> PlatformXRSystem {
+#if USE(OPENXR)
+    CreateLayerProjection(uint32_t width, uint32_t height, bool alpha)
+    SubmitFrame(Vector<WebKit::XRDeviceLayer> layers)
+#endif
+#if !USE(OPENXR)
```
Any reason to not use `#else` here? This looks more readable to me:

```suggestion
#else
```

svillar replied:
Because that is not accepted in messages.in files

aperezdc commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -0,0 +1,105 @@
+        m_projectionViews[i].subImage.imageRect.extent = { viewport.width(), viewport.height() };
+    }
+
+    m_layerProjection.space = space;
+    m_layerProjection.viewCount = m_projectionViews.size();
+    m_layerProjection.views = m_projectionViews.span().data();
```
For a moment I thought the `.layerFlags` member was not being set, it took me a bit to realize that it had already been initialized above. Maybe for clarity it would make sense to move the assignment of `m_layerProjection.layerFlags` here so all the the changes to the same struct are kept together.

aperezdc commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -207,19 +229,34 @@ void OpenXRCoordinator::scheduleAnimationFrame(WebPageProxy& page, std::optional
 {
-    Vector<XrCompositionLayerBaseHeader*> layers;
+    WebCore::GLContext::ScopedGLContextCurrent scopedContext(*m_glContext);
+    Vector<const XrCompositionLayerBaseHeader*> frameEndLayers;
```
If we have an idea of the typical amount of layers we would get here, and it's low, it may be good to pass a second parameter to the `Vector` template to use (faster) stack-allocated space for the vector.

(Could be premature optimization, so feel free to ignore this comment.)

svillar replied:
so far we'll only have 1 as WebXR layers spec is not implemented in WebKit, so we could add it

aperezdc commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -426,6 +457,30 @@ void OpenXRCoordinator::createSessionIfNeeded()
 
+void OpenXRCoordinator::cleanupSessionAndAssociatedResources()
+{
+    if (m_localSpace != XR_NULL_HANDLE) {
+        CHECK_XRCMD(xrDestroySpace(m_localSpace));
+        m_localSpace = XR_NULL_HANDLE;
+    }
```
Here's an idea&mdash;for the future, no need to apply any change in this PR: maybe at some point we would want to add a `std::unique_ptr` specialization for OpenXR pointer types, the same way we have e.g. `GUniquePtr`. That would make this kind of code unnecessary, leaving the task to the smart pointer specialization.

svillar replied:
Looks like a nice idea for a future refactoring indeed. The challenge is that most of them are just opaque handlers and you need to call the specific destroy function for each of them, perhaps too much complexity.

carlosgcampos commented on Source/WebKit/UIProcess/XR/PlatformXRCoordinator.h:
```diff
@@ -29,6 +29,9 @@
 
 #include "XRDeviceIdentifier.h"
 #include "XRDeviceInfo.h"
+#if USE(OPENXR)
```
The header is already guarded.

svillar replied:
We need that anyway because the file is not included in !OPENXR builds

carlosgcampos commented on Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
```diff
@@ -249,7 +253,11 @@ void PlatformXRSystem::submitFrame(IPC::Connection& connection)
         return;
 
     if (auto* xrCoordinator = PlatformXRSystem::xrCoordinator())
+#if USE(OPENXR)
+        xrCoordinator->submitFrame(*page, WTFMove(layers));
+#else
         xrCoordinator->submitFrame(*page);
+#endif
```
Add { } to the if now.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.h:
```diff
@@ -0,0 +1,68 @@
+public:
+    static std::unique_ptr<OpenXRLayerProjection> create(XrInstance, XrSession, uint32_t width, uint32_t height, int64_t format, uint32_t sampleCount);
+private:
+    OpenXRLayerProjection(UniqueRef<OpenXRSwapchain>&&);
```
```suggestion
    explicit OpenXRLayerProjection(UniqueRef<OpenXRSwapchain>&&);
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.h:
```diff
@@ -0,0 +1,66 @@
+#if ENABLE(WEBXR) && USE(OPENXR)
+
+#include "OpenXRUtils.h"
+
```
```suggestion
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.h:
```diff
@@ -0,0 +1,66 @@
+
+#include <WebCore/GraphicsTypesGL.h>
+#include <WebCore/IntSize.h>
+typedef void* EGLDisplay;
+typedef void* EGLContext;
+typedef void* EGLConfig;
+typedef unsigned EGLenum;
```
Move these after the includes block

carlosgcampos replied:
I don't see these used in this header, though

svillar replied:
They need to be there because they're required by openxr_platform.h which is just bellow them

carlosgcampos replied:
Why aren't they in that header then?

svillar replied:
Because that header is multiplatform and multi-graphics libraries. It's full of #if's that define various graphics types depending on which graphics library you are using.

carlosgcampos replied:
I see, we need to find a solution (in a different patch). In cpp files we just need to make sure we include epoxy or egl platform headers before. In header files it's more complicated. Could we avoid including openxr_platform.h in headers?

svillar replied:
I guess it could be  done. Perhaps adding private opaque objects that are defined in the sources...

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.h:
```diff
@@ -0,0 +1,66 @@
+    PlatformGLObject m_acquiredTexture { 0 };
+};
+
+
```
```suggestion
```

carlosgcampos commented on Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
```diff
@@ -103,15 +103,37 @@ void PlatformXRSystemProxy::requestFrame(std::optional<PlatformXR::RequestData>&
+    UNUSED_PARAM(height);
+    UNUSED_PARAM(alpha);
+#endif
     return PlatformXRCoordinator::defaultLayerHandle();
```
This is confusing, shouldn't the IPC message return the layer handle?

svillar replied:
It's unused so far, but yeah I agree that it makes more sense, and it'll be better for future use

svillar committed:
[WebXR][OpenXR] Fill in FrameData information
https://bugs.webkit.org/show_bug.cgi?id=295881

Reviewed by Carlos Garcia Campos.

XR information coming from XR devices is retrieved in the UI process
using OpenXR API. That info is then later transferred to the Web
Process to be consumed by the WebXR API. A struct called FrameData
is used to carry that information.

In this commit we're among other things collecting & sending:
- Views information: fov, pose
- Tracking information availability
- Floor location (transform from local space)
- Layer data (so far only 1 projection layer)

In the WebProcess the submitFrame() method receives a vector
of objects containing information about the layers. The WebXR
layers spec is still not available in WebKit but even without
that there is still at least one layer, the base layer.

That info was not passed to the UIProcess in submitFrame() because
it wasn't used by the ARKit implementation. However it's needed
for the OpenXR one because we get there the viewport information
that we need to create the stereo views (we use a single projection
layer that is split in as many parts as views we have, typically 2).

Serializing PlatformXR::Device::Layer proved to be really complex
as serializers do not deal well with having multiple nested namespaces
for a given type. That's why we had to "clone" that structure in
another one hanging directly from the WebKit namespace.

Also a new message had to be added as well to ask the UIProcess to
create the projection layer with the data coming from WebXR.

One of the most important pieces of information to send to the Web Process
is the texture were the WebGL rendering should happen to be later submitted
to the OpenXR compositor. This commit implements the required machinery
to acquire and release those textures coming from OpenXR. However that
cannot be directly used by WebGL, we need to eventually implement a similar
mechanism as the ExternalTexture used by Apple ports. That's why no
rendering is expected to actually happen in the texture provided by OpenXR.

Last but not least, when adding the swapchains code we detected 2 important
issues in the current implementation:
- OpenXR resources were not properly freed upon session exiting
- we were incorrectly running multiple sequences of xrBeginFrame->xrEndFrame
in parallel. The confusion arose from an incorrect interpretation of what
xrWaitFrame does. The current code assumed that xrWaitFrame would wait until the
previous xrEndFrame is finished. But that's wrong. As the specs say
"A subsequent xrWaitFrame call must block until the previous frame has been begun
with xrBeginFrame and must unblock independently of the corresponding call
to xrEndFrame". That's why we had to add some additional synchronization
using binary semaphores, basically identical to what ARKit implementation does.

* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/Shared/XR/XRDeviceLayer.h: Copied from Source/WebKit/UIProcess/XR/openxr/PlatformXRSystemOpenXR.cpp.
* Source/WebKit/Shared/XR/XRDeviceProxy.cpp:
(WebKit::XRDeviceProxy::initializeTrackingAndRendering):
(WebKit::XRDeviceProxy::submitFrame):
* Source/WebKit/Shared/XR/XRSystem.serialization.in:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/XR/PlatformXRCoordinator.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::submitFrame):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.messages.in:
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp: Added.
(WebKit::OpenXRLayerProjection::create):
(WebKit::OpenXRLayerProjection::OpenXRLayerProjection):
(WebKit::OpenXRLayerProjection::startFrame):
(WebKit::OpenXRLayerProjection::endFrame):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.h: Added.
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.cpp: Added.
(WebKit::OpenXRSwapchain::create):
(WebKit::OpenXRSwapchain::OpenXRSwapchain):
(WebKit::OpenXRSwapchain::~OpenXRSwapchain):
(WebKit::OpenXRSwapchain::acquireImage):
(WebKit::OpenXRSwapchain::releaseImage):
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.h: Added.
(WebKit::OpenXRSwapchain::swapchain const):
(WebKit::OpenXRSwapchain::width const):
(WebKit::OpenXRSwapchain::height const):
(WebKit::OpenXRSwapchain::size const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRUtils.h:
(WebKit::XrIdentityPose):
(WebKit::XrPosefToPose):
(WebKit::XrViewToView):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::~OpenXRCoordinator):
(WebKit::OpenXRCoordinator::getPrimaryDeviceInfo):
(WebKit::OpenXRCoordinator::createLayerProjection):
(WebKit::OpenXRCoordinator::endSessionIfExists):
(WebKit::OpenXRCoordinator::submitFrameInternal):
(WebKit::OpenXRCoordinator::submitFrame):
(WebKit::OpenXRCoordinator::collectViewConfigurations):
(WebKit::OpenXRCoordinator::initializeBlendModes):
(WebKit::OpenXRCoordinator::cleanupSessionAndAssociatedResources):
(WebKit::OpenXRCoordinator::handleSessionStateChange):
(WebKit::OpenXRCoordinator::populateFrameData):
(WebKit::OpenXRCoordinator::createReferenceSpacesIfNeeded):
(WebKit::OpenXRCoordinator::renderLoop):
(WebKit::OpenXRCoordinator::recommendedResolution const): Deleted.
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:
* Source/WebKit/UIProcess/XR/openxr/PlatformXRSystemOpenXR.cpp:
(WebKit::PlatformXRSystem::createLayerProjection):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.cpp:
(WebKit::PlatformXRSystemProxy::createLayerProjection):
(WebKit::PlatformXRSystemProxy::submitFrame):
* Source/WebKit/WebProcess/XR/PlatformXRSystemProxy.h:

Canonical link: https://commits.webkit.org/297780@main
